JuliaStream.jl/src/Stream.jl

```diff
@@ -0,0 +1,291 @@
+    elseif benchmark == Nstream
+      gold_a += gold_b + data.scalar * gold_c
+    else
+      error("Unknown benchmark", benchmark)
+    end
+  end
+
+  tolerance = eps(T) * 100
+  function validate_xs(name::String, xs::AbstractArray{T}, from::T)
+    error = (map(x -> abs(x - from), xs) |> sum) / length(xs)
+    failed = error > tolerance
+    if failed
+      println("Validation failed on $name. Average error $error")
+    end
+    !failed
+  end
+  a_valid = validate_xs("a", data.a, gold_a)
+  b_valid = validate_xs("b", data.b, gold_b)
+  c_valid = validate_xs("c", data.c, gold_c)
+  dot_valid =
+    dot !== nothing ?
+    begin
+      gold_sum = gold_a * gold_b * data.size
+      error = abs((dot - gold_sum) / gold_sum)
+      failed = error > 1.0e-8
+      if failed
+        println(
+          "Validation failed on sum. Error $error \nSum was $dot but should be $gold_sum",
+        )
+      end
+      !failed
+    end : true
+
+  a_valid && b_valid && c_valid && dot_valid
+end
+
+@with_kw mutable struct Config
+  list::Bool = false
+  device::Int = 1
+  numtimes::Int = 100
+  arraysize::Int = 33554432
+  float::Bool = false
+  triad_only::Bool = false
+  nstream_only::Bool = false
+  csv::Bool = false
+  mibibytes::Bool = false
+end
+
+function parse_options(given::Config)
+  s = ArgParseSettings()
+  @add_arg_table s begin
+    "--list"
+    help = "List available devices"
+    action = :store_true
+    "--device", "-d"
+    help = "Select device at DEVICE, NOTE: Julia is 1-indexed"
+    arg_type = Int
+    default = given.device
+    "--numtimes", "-n"
+    help = "Run the test NUMTIMES times (NUM >= 2)"
+    arg_type = Int
+    default = given.numtimes
+    "--arraysize", "-s"
+    help = "Use ARRAYSIZE elements in the array"
+    arg_type = Int
+    default = given.arraysize
+    "--float"
+    help = "Use floats (rather than doubles)"
+    action = :store_true
+    "--triad_only"
+    help = "Only run triad"
+    action = :store_true
+    "--nstream_only"
+    help = "Only run nstream"
+    action = :store_true
+    "--csv"
+    help = "Output as csv table"
+    action = :store_true
+    "--mibibytes"
+    help = "Use MiB=2^20 for bandwidth calculation (default MB=10^6)"
+    action = :store_true
+  end
+  args = parse_args(s)
+  # surely there's a better way than doing this:
+  for (arg, val) in args
+    setproperty!(given, Symbol(arg), val)
+  end
+end
+
+const DefaultInit = (0.1, 0.2, 0.0)
+const DefaultScalar = 0.4
+const Version = "3.4.0"
+
+function main()
+
+  config::Config = Config()
+  parse_options(config)
+
+  if config.list
+    ds = devices()
+    for (i, device) in enumerate(ds)
+      println("[$i] $(device)")
+    end
+    exit(0)
+  end
+
+  ds = devices()
+  if config.device < 1 || config.device > length(ds)
+    error(
+      "Device $(config.device) out of range (1..$(length(ds))), NOTE: Julia is 1-indexed",
+    )
+  end
+
+  if config.float
+    type = Float32
+  else
+    type = Float64
+  end
```

Comment:
For the record, you can also do
```suggestion
  type = if config.float
    Float32
  else
    Float64
  end
```
or even
```suggestion
  type = config.float ? Float32 : Float64
```
if you want to be even more concise.  Also in the `if` below you can save a couple of `benchmark =` in the same way